Source/WebCore/workers/service/background-fetch/BackgroundFetchManager.cpp

```diff
@@ -28,31 +28,147 @@
+BackgroundFetchManager::BackgroundFetchManager(ServiceWorkerRegistration& registration)
+    : m_identifier(registration.identifier())
 {
 }
 
 BackgroundFetchManager::~BackgroundFetchManager()
 {
 }
 
-void BackgroundFetchManager::fetch(ScriptExecutionContext&, const String&, Requests&&, BackgroundFetchOptions&&, DOMPromiseDeferred<IDLInterface<BackgroundFetchRegistration>>&& promise)
+static ExceptionOr<Vector<Ref<FetchRequest>>> buildBackgroundFetchRequests(ScriptExecutionContext& context, BackgroundFetchManager::Requests&& backgroundFetchRequests)
 {
-    promise.reject(Exception { NotSupportedError });
+    if (std::holds_alternative<RefPtr<FetchRequest>>(backgroundFetchRequests)) {
+        auto result = FetchRequest::create(context, std::get<RefPtr<FetchRequest>>(backgroundFetchRequests).releaseNonNull(), { });
+        if (result.hasException())
+            return result.releaseException();
+        if (result.returnValue()->mode() == FetchOptions::Mode::NoCors)
+            return Exception { TypeError, "Request has no-cors mode"_s };
+        return Vector<Ref<FetchRequest>> { result.releaseReturnValue() };
+    }
+
+    if (std::holds_alternative<String>(backgroundFetchRequests)) {
+        auto result = FetchRequest::create(context, WTFMove(std::get<String>(backgroundFetchRequests)), { });
+        if (result.hasException())
+            return result.releaseException();
+        return Vector<Ref<FetchRequest>> { result.releaseReturnValue() };
+    }
+
+    auto& requestInfos = std::get<Vector<BackgroundFetchManager::RequestInfo>>(backgroundFetchRequests);
+    std::optional<Exception> exception;
+    Vector<Ref<FetchRequest>> requests;
+    requests.reserveInitialCapacity(requestInfos.size());
+    for (auto& requestInfo : requestInfos) {
+        auto result = FetchRequest::create(context, WTFMove(requestInfo), { });
+        if (result.hasException())
+            return result.releaseException();
+        if (result.returnValue()->mode() == FetchOptions::Mode::NoCors)
+            return Exception { TypeError, "Request has no-cors mode"_s };
+
+        // FIXME: Add support for readable stream bodies
+        if (result.returnValue()->isReadableStreamBody())
+            return Exception { NotSupportedError, "ReadableStream uploading is not supported"_s };
+
+        requests.uncheckedAppend(result.releaseReturnValue());
+    }
+    return requests;
+}
+
+Ref<BackgroundFetchRegistration> BackgroundFetchManager::backgroundFetchRegistrationInstance(ScriptExecutionContext& context, BackgroundFetchInformation&& data)
+{
+    auto identifier = data.identifier;
+    auto result = m_backgroundFetchRegistrations.ensure(identifier, [data = WTFMove(data), &context]() mutable {
+        return BackgroundFetchRegistration::create(context, WTFMove(data));
+    });
+
+    auto registration = result.iterator->value;
+    if (!result.isNewEntry)
+        registration->updateInformation(data);
```

Comment:
use-after-move of `data`?

Reply:
Not really, since the lambda is not called if there is already an entry. I'll think of a better way to write it.

Reply:
May look clearer to capture `&data` instead.

Reply:
or simply `[&]`